src/eachobs.jl

```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
 end
 ```
 """
-function eachobs(data; buffer = false, batchsize::Int = -1, partial::Bool =true)
+function eachobs(
+        data;
+        buffer = false,
+        parallel = false,
+        shuffle = false,
+        batchsize::Int = -1,
+        partial::Bool = true,
+        executor = _default_executor())
     if batchsize > 0
         data = BatchView(data; batchsize, partial)
     end
-    if buffer === false 
-        gen = (getobs(data, i) for i in 1:numobs(data))
+
+    iter = if parallel
+        eachobsparallel(data; buffer, executor)
     else
-        if buffer === true && numobs(data) > 0
-            buffer = getobs(data, 1)
+        if buffer === false
+            EachObs(data)
+        elseif buffer === true
+            EachObsBuffer(data)
+        else
+            EachObsBuffer(data, buffer)
         end
-        gen = (getobs!(buffer, data, i) for i in 1:numobs(data))
     end
-    return gen
+
+    if shuffle
+        iter = ReshuffleIter(iter)
+    end
+    return iter
+end
+
+
+# Internal
+
+"""
+    EachObs(data)
+
+Create an iterator over observations in data container `data`.
+
+This is an internal function. Use `eachobs(data)` instead.
+"""
+struct EachObs{T}
+    data::T
 end
+Base.length(iter::EachObs) = numobs(iter.data)
+Base.eltype(iter::EachObs) = eltype(iter.data)
 
+function Base.iterate(iter::EachObs, i::Int = 1)
+    i > numobs(iter) && return nothing
+    return getobs(iter.data, i), i+1
+end
+
+
+"""
+    EachObsBuffer(data)
+
+Create a buffered iterator over observations in data container `data`.
+Buffering only works if [`getobs!`](@ref) is implemented for `data`.
+
+This is an internal function. Use `eachobs(data, buffer = true)` instead.
+"""
+struct EachObsBuffer{T, B}
+    data::T
+    buffer::B
+end
+EachObsBuffer(data) = EachObsBuffer(data, getobs(data, 1))
+Base.length(iter::EachObsBuffer) = numobs(iter.data)
+Base.eltype(iter::EachObsBuffer) = eltype(iter.data)
+
+function Base.iterate(iter::EachObsBuffer)
+    obs = getobs!(iter.buffer, iter.data, 1)
+    return obs, (obs, 2)
+end
+
+function Base.iterate(iter::EachObsBuffer, (buffer, i))
+    i > numobs(iter) && return nothing
+    return getobs!(buffer, iter.data, i), (buffer, i+1)
```

Comment:
Since `getobs!(nothing, x, i)` is equivalent to `getobs(x, i)`, we could use a unique struct
```julia
struct EachObs
  data
  buffer
end
```
for both buffered and unbuffered iterations

Reply:
Can do, though I think we would lose the ability to turn off buffering if we just used the same struct for both. So unless I am missing something we would still need to store a flag as to whether buffering is turned on.

If we're doing that, it may make sense to go and add a `parallel` flag as well that is checked when starting to iterate?

Reply:
> Can do, though I think we would lose the ability to turn off buffering 

Not exactly sure of what you mean here. A call to `eachobs(data, buffer=false)` will produce a `EachObs(data, nothing)` object for which we can `getobs!(iter.buffer, iter.data, i)` falling back to `getobs(iter.data, i)`

Reply:
I agree with Carlo about consolidating as much as possible. If we add `buffer` to the regular `EachObs`, then `obs = getobs!(buffer, data, i)` will just use the non-buffered `getobs` when `buffer == nothing` (this is Carlo's point, I think). We shouldn't need a flag cause we just used the stored `iter.buffer` which is `nothing`.

The case of `parallel` seems like it would be harder to consolidate the two structs, so I would prefer to return the appropriate one in `eachobs` like you are now.

Reply:
Sorry a bit of a double-post refresh issue!